### PR TITLE
Add rate limiting of /api/annotations

### PR DESCRIPTION
**Pull request by hmstepanek**
Set an overall request rate limit per user. This will establish "fairness" between users and prevent one user from DDOSing other users and also issuing too many requests that the service simply can't handle. Note the difference between this and Cloudflare is the rate limiting is done per authtoken fallback to ip rather than per ip. 

Rate limit endpoints that are known to cause issues if over-requested
on a per user basis (use auth token to identify an individual user fallback
to using the ip address). Use the authorization token instead of ip address 
since users from a university may have the same ip address and the bulk 
of users in h day-to-day are students. 

Our current traffic allocation is as follows:
 - 60% badge
 - 22% search
 - <1% create

We currently have 12 gunicorn workers which means we have 2.7 workers for
handling search requests and .12 workers for handling create annotation requests.

/api/annotations are held in a request queue at the nginx level during bursty
traffic and forwarded to the server at a set rate. This prevents the server from
being overwhelmed with an influx of db intensive operations which will cause db
slowdowns and result in a larger than expected amount of gunicorn worker time 
spent processing create requests. See [incident report](https://docs.google.com/document/d/1AIbl0BMw9FP0cza1RAFNIvzo2ZbVpfIYAqlSxnAtJG8/edit#heading=h.3de8xgxrwmfo) for details. At peak load
this particular bot was running ~72rpm.

The large blue spikes in the graph below are from a bot that runs w/o crashing
our app today. The tricky part is this doesn't crash our app because it runs during
low traffic times, if this same traffic ran earlier in the day it would overwhelm the
server.
image

See https://www.nginx.com/blog/rate-limiting-nginx/ for details.

### CODE REVIEW

**Comment by hmstepanek** [comment]
[slow create transactions](https://rpm.newrelic.com/accounts/1385283/applications/22688631/transactions)

**Comment by hmstepanek** [comment]
[priority](http://nginx.org/en/docs/http/ngx_http_core_module.html#location)

**Comment by robertknight** [comment]
Upon further testing I've realized there is an additional problem with the current setup.

When a request is processed, the following happens:

1. nginx iterates over all the limits defined using `limit_req` directives
2. For each limit, the request's key is computed according to the associated shared-memory zone
3. The stored state associated with that key is fetched from the zone. Note that this state is shared _by all locations which have rate limiting directives that share the same zone._ This state includes a counter which keeps track of the excess requests above the zone's rate limit.
4. The excess is updated depending on the time since the previous request.
5. The excess is compared to the burst value for the current limit being checked. If it is higher, the request is rejected.
6. The updated excess is then stored back into the zone, associated with the key.

What this means is that if I do the following:

1. Make 50 requests to `/assets/images/icons/logo.svg` route within a couple of seconds.
2. Make 1 request to `/_status` a second later.

Then step (1) succeeds because the `/assets/...` route has a high burst limit, but step (2) fails because step (1) stored an excess value (~50) which is higher than the burst limit for `/_status` (44) and not enough time has expired for the excess to be "leaked".

Gist to reproduce: https://gist.github.com/robertknight/4f2f0b51424ebd65171179c496656ca6

If we don't want requests for one location to count against the limit for another location then we'll have to define separate zones for accounting purposes.

File: conf/nginx.conf [hunk 1]
```diff
@@ -56,6 +68,60 @@ http {
...
 
     location / {
+      # A bot may burst up to 50rpm and the client issues 5 requests upon
+      # loading the sidebar. Assume a max request rate of 15 rps in a burst.
+      # This means the queue size would be 14 requests.
+      # Allow a user to sustain the max bursty request rate for 3 seconds.
+	  # This would mean they can request up to 45 requests in one second but
+      # only 1 new request each second after that.
+      limit_req zone=user_1rps_limit burst=44 nodelay;
+      limit_req_status 429;
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply conf/nginx.conf]: I moved the proxy setting to the top of the block. I also re-ordered the rate limit settings so they follow if-else ordering (meaning I moved the specific locations above the catchall (previously on line 77) for all other api requests.

File: conf/nginx.conf [hunk 2]
```diff
@@ -56,6 +68,57 @@ http {
...
+      # The 95th percentile time for an asset is .013s.
+      # The maximum burst of requests from a single page
+      # https://hypothes.is/docs/help is 20 requests.
+      # <1% of worker time is spent handling these requests.
+      # The maximum expected request rate is 25rpm.
+      # Assume in frustration the user hammers on the refresh
+      # button 7 times in a row. Worst case this results in
+      # a burst of 140 requests.
+      location /assets {
+        limit_req zone=user_1rps_limit burst=139 nodelay;
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply conf/nginx.conf]: I fixed this as you suggested-by adding different zones.

**Comment by robertknight** [comment]
In testing the JSON response I realized that one slightly tricky aspect of rate limiting for browser-based clients is that they will send _two_ HTTP requests instead of one, if the request [requires preflighting](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Preflighted_requests).

1. An initial "preflight" `OPTIONS` request to check whether the client is allowed to make a request with a particular method to a particular endpoint.
2. If (1) responds with an HTTP 200, and the Access-Control-Allow-Method and Access-Control-Allow-Origin headers on the response permit the request, then the actual request itself gets made.

When rate limiting kicks in, both requests will often end up getting rate limited. As a result instead of getting a 429 error which it can parse, the client gets a super-generic (for security reasons) "Failed to fetch" error because the preflight request failed.

**Comment by hmstepanek** [comment]
Detailed calculation comments have been replaced with a general high level description of how the number was chosen. The detailed calculations are in the body of the commit message that removes the detailed comments so there is a record of how that number was calculated in detail in the commit history.

**Comment by hmstepanek** [comment]
I'll merge this Friday morning so I have a whole day to monitor it in qa and prod.

**Comment by dwhly** [comment]
If rate limiting is triggered, is there a way to know?  Or a way to track the metrics of limiting and to whom?

**Comment by hmstepanek** [comment]
Rate limiting metrics have been added in aws cloudwatch.

**Pull request merged**